sacred/observers/s3_observer.py

```diff
@@ -0,0 +1,328 @@
+    for char in bucket_name:
+        if char.isdigit():
+            continue
+        if char.islower():
+            continue
+        if char == '-':
+            continue
+        return False
+    try:
+        # If a name is a valid IP address, it cannot be a bucket name
+        socket.inet_aton(bucket_name)
+    except socket.error:
+        return True
+
+
+class S3FileObserver(RunObserver):
+    VERSION = 'S3FileObserver-0.1.0'
+
+    @classmethod
+    def create(cls, bucket, basedir, resource_dir=None, source_dir=None,
+               priority=DEFAULT_S3_PRIORITY):
+        resource_dir = resource_dir or os.path.join(basedir, '_resources')
+        source_dir = source_dir or os.path.join(basedir, '_sources')
+
+        return cls(bucket, basedir, resource_dir, source_dir, priority)
+
+    def __init__(self, bucket, basedir, resource_dir, source_dir,
+                 priority=DEFAULT_S3_PRIORITY):
+        if not _is_valid_bucket(bucket):
+            raise ValueError("Your chosen bucket name does not follow AWS "
+                             "bucket naming rules")
+
+        self.basedir = basedir
+        self.bucket = bucket
+        # Keeping the convention of referring to locations in S3 as `dir`
+        # because that is a useful mental model and there isn't a better word
+        self.resource_dir = resource_dir
+        self.source_dir = source_dir
+        self.priority = priority
+        self.dir = None
+        self.run_entry = None
+        self.config = None
+        self.info = None
+        self.cout = ""
+        self.cout_write_cursor = 0
+        self.s3 = boto3.resource('s3')
+        self.saved_metrics = {}
+
+    def _objects_exist_in_dir(self, prefix):
+        # This should be run after you've confirmed the bucket
+        # exists, and will error out if it does not exist
+
+        bucket = self.s3.Bucket(self.bucket)
+        all_keys = [el.key for el in bucket.objects.filter(Prefix=prefix)]
+        return len(all_keys) > 0
+
+    def _list_s3_subdirs(self, prefix=None):
+        if prefix is None:
+            prefix = self.basedir
+        try:
+            bucket = self.s3.Bucket(self.bucket)
+            all_keys = [el.key for el in bucket.objects.filter(Prefix=prefix)]
+        except ClientError as er:
+            if er.response['Error']['Code'] == 'NoSuchBucket':
+                return None
+            else:
+                raise ClientError(er.response['Error']['Code'])
+
+        subdir_match = r'{prefix}\/(.*)\/'.format(prefix=prefix)
+        subdirs = []
+        for key in all_keys:
+            match_obj = re.match(subdir_match, key)
+            if match_obj is None:
+                continue
+            else:
+                subdirs.append(match_obj.groups()[0])
+        distinct_subdirs = set(subdirs)
+        return list(distinct_subdirs)
+
+    def _create_bucket(self):
+        session = boto3.session.Session()
+        current_region = session.region_name or 'us-west-2'
```

Comment:
Not sure if using a default here is good. Shouldn't the client take care of specifying the region?

Reply:
Sure, that seems reasonable. I could change it so that you pass in region as an optional parameter, and then when the Observer is created, it checks whether you have a region_name defined in your session, and if you don't, and if you haven't passed in a region to the constructor, it will fail to be constructed, because it doesn't have a valid region set.